lib/hijack/wrap_webapp.js

```diff
@@ -78,69 +46,73 @@ export async function wrapWebApp () {
-        }
+    Kadira.tracer.event(trace, EventType.Start, {
+      url: req.url,
+      method: req.method,
+      headers: JSON.stringify(headers),
+    });
 
-        const isJson = req.headers['content-type'] === 'application/json';
-        const hasSmallBody = req.headers['content-length'] > 0 && req.headers['content-length'] < MAX_BODY_SIZE;
-
-        // Check after all middleware have run to see if any of them
-        // set req.body
-        // Technically bodies can be used with any method, but since many load balancers and
-        // other software only support bodies for POST requests, we are
-        // not recording the body for other methods.
-        if (req.method === 'POST' && req.body && isJson && hasSmallBody) {
-          try {
-            let body = JSON.stringify(req.body);
-
-            // Check the body size again in case it is much
-            // larger than what was in the content-length header
-            if (body.length < MAX_STRINGIFIED_BODY_SIZE) {
-              trace.events[0].data.body = body;
-            }
-          } catch (e) {
-          // It is okay if this fails
+    req.__kadiraInfo = { trace };
+
+    res.on('finish', () => {
+      Kadira.tracer.endLastEvent(trace);
+
+      if (req.__kadiraInfo.isStatic) {
+        trace.name = `${req.method}-<static file>`;
+      } else if (req.__kadiraInfo.suggestedRouteName) {
+        trace.name = `${req.method}-${req.__kadiraInfo.suggestedRouteName}`;
+      } else if (req.__kadiraInfo.isAppRoute) {
+        trace.name = `${req.method}-<app>`;
+      }
+
+      const isJson = req.headers['content-type'] === 'application/json';
+      const hasSmallBody = req.headers['content-length'] > 0 && req.headers['content-length'] < MAX_BODY_SIZE;
+
+      // Check after all middleware have run to see if any of them
+      // set req.body
+      // Technically bodies can be used with any method, but since many load balancers and
+      // other software only support bodies for POST requests, we are
+      // not recording the body for other methods.
+      if (req.method === 'POST' && req.body && isJson && hasSmallBody) {
+        try {
+          let body = JSON.stringify(req.body);
+
+          // Check the body size again in case it is much
+          // larger than what was in the content-length header
+          if (body.length < MAX_STRINGIFIED_BODY_SIZE) {
+            trace.events[0].data.body = body;
           }
+        } catch (e) {
+          // It is okay if this fails
         }
+      }
 
-        // TODO: record status code
-        Kadira.tracer.event(trace, 'complete');
-        let built = Kadira.tracer.buildTrace(trace);
-        Kadira.models.http.processRequest(built, req, res);
-      });
+      // TODO: record status code
+      Kadira.tracer.event(trace, EventType.Complete);
+      let built = Kadira.tracer.buildTrace(trace);
+      Kadira.models.http.processRequest(built, req, res);
+    });
 
-      next();
-    }
-  });
+    next();
```

Comment:
In theory, we should be able to simply wrap the `next` function, and ALS will preserve the kadira info through all of the middleware.